lib/trino-hdfs/src/main/java/io/trino/hdfs/TrinoFileSystemCache.java

```diff
@@ -102,55 +96,45 @@ public FileSystem getUnique(URI uri, Configuration conf)
     @VisibleForTesting
     int getCacheSize()
     {
-        return map.size();
+        return cache.size();
     }
 
-    private synchronized FileSystem getInternal(URI uri, Configuration conf, long unique)
+    private FileSystem getInternal(URI uri, Configuration conf, long unique)
             throws IOException
     {
         UserGroupInformation userGroupInformation = UserGroupInformation.getCurrentUser();
         FileSystemKey key = createFileSystemKey(uri, userGroupInformation, unique);
         Set<?> privateCredentials = getPrivateCredentials(userGroupInformation);
 
-        FileSystemHolder fileSystemHolder = map.get(key);
-        if (fileSystemHolder == null) {
-            int maxSize = conf.getInt("fs.cache.max-size", 1000);
-            if (map.size() >= maxSize) {
-                stats.newGetCallFailed();
-                throw new IOException(format("FileSystem max cache size has been reached: %s", maxSize));
-            }
-            try {
-                FileSystem fileSystem = createFileSystem(uri, conf);
-                fileSystemHolder = new FileSystemHolder(fileSystem, privateCredentials);
-                map.put(key, fileSystemHolder);
-            }
-            catch (IOException e) {
-                stats.newGetCallFailed();
-                throw e;
-            }
+        int maxSize = conf.getInt("fs.cache.max-size", 1000);
+        FileSystemHolder fileSystemHolder;
+        try {
+            fileSystemHolder = cache.compute(key, (k, currFileSystemHolder) -> {
+                if (currFileSystemHolder == null) {
+                    if (cacheSize.getAndUpdate(curr -> curr < maxSize ? (curr + 1) : curr) >= maxSize) {
+                        throw new RuntimeException(
+                                new IOException(format("FileSystem max cache size has been reached: %s", maxSize)));
+                    }
+                    return new FileSystemHolder(uri, conf, privateCredentials);
+                }
+                else {
+                    // Update file system instance when credentials change.
+                    if (currFileSystemHolder.credentialsChanged(uri, conf, privateCredentials)) {
+                        return new FileSystemHolder(uri, conf, privateCredentials);
+                    }
+                    else {
+                        return currFileSystemHolder;
+                    }
+                }
+            });
+
+            fileSystemHolder.createFileSystemOnce();
```

Comment:
Can we add a comment here why this is outside of the cache compute? Seems like that's an important piece of why this works